autodispose-kotlin/src/test/kotlin/com/uber/autodispose/kotlin/AutoDisposeKotlinTest.kt

```diff
@@ -0,0 +1,545 @@
+/*
+ * Copyright (c) 2017. Uber Technologies
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.uber.autodispose.kotlin
+
+import com.uber.autodispose.LifecycleEndedException
+import com.uber.autodispose.LifecycleNotStartedException
+import com.uber.autodispose.LifecycleScopeProvider
+import com.uber.autodispose.ScopeProvider
+import com.uber.autodispose.kotlin.LifecycleEvent.End
+import com.uber.autodispose.kotlin.LifecycleEvent.Start
+import io.reactivex.BackpressureStrategy.ERROR
+import io.reactivex.Completable
+import io.reactivex.Flowable
+import io.reactivex.Maybe
+import io.reactivex.Observable
+import io.reactivex.Single
+import io.reactivex.functions.Function
+import io.reactivex.observers.TestObserver
+import io.reactivex.subjects.BehaviorSubject
+import io.reactivex.subjects.CompletableSubject
+import io.reactivex.subjects.MaybeSubject
+import io.reactivex.subjects.PublishSubject
+import io.reactivex.subjects.SingleSubject
+import io.reactivex.subscribers.TestSubscriber
+import org.junit.Test
+
+sealed class LifecycleEvent {
+  class Start: LifecycleEvent()
+  class End: LifecycleEvent()
+}
+
+class AutoDisposeKotlinTest {
+
+  val o = TestObserver<String>()
+  val s = TestSubscriber<String>()
+  val scopeMaybe = MaybeSubject.create<Any>()
+  val scopeProvider = ScopeProvider { scopeMaybe }
+  val lifecycleEvents = BehaviorSubject.create<LifecycleEvent>()
+  val lifecycle = object : LifecycleScopeProvider<LifecycleEvent> {
+    override fun lifecycle(): Observable<LifecycleEvent> {
+      return lifecycleEvents
+    }
+
+    override fun peekLifecycle(): LifecycleEvent? {
+      return lifecycleEvents.value
+    }
+
+    override fun correspondingEvents(): Function<LifecycleEvent, LifecycleEvent> {
+      return Function { event ->
+        when (event) {
+          is Start -> End()
+          is End -> throw LifecycleEndedException()
+        }
+      }
+    }
+  }
+
+  @Test fun observable_maybeNormalCompletion() {
+    Observable.just("Hello")
+        .autoDisposeWith(scopeMaybe)
+        .subscribe(o)
+
+    o.assertValue { v -> v == "Hello" }
```

Comment:
Good point - 2e047aa

Forgot this makes `when` blocks with instance checks really pretty.

```kotlin
return Function {
  when (it) {
    is Start -> End()
    is End -> throw LifecycleEndedException()
  }
}
```